### ISSUE TITLE
Uncaught SyntaxError: Unexpected token import

### ISSUE THREAD

**Issue by klimashkin** [webpack/webpack#4857]
**Bug**

**What is the current behavior?**
In case of development environment such code
```
if (process.env.NODE_ENV === 'testing') {
  import('somemodule');
}
```
is transformed by webpack `2.5.1` into this
```
if (false) {
  import('somemodule');
}
```
which leads to `Uncaught SyntaxError: Unexpected token import` error in the latest Chrome.

Node.JS 7.10.0, babel 6.24.1 has `syntax-dynamic-import` plugin

**Comment by anlexN** [comment]
![dev1](https://cloud.githubusercontent.com/assets/8686739/26191792/e0e1d5ac-3be1-11e7-8a9f-411740aac966.PNG)
![dev2](https://cloud.githubusercontent.com/assets/8686739/26191789/e0ce53b0-3be1-11e7-8632-03fecc3d72b4.PNG)
![dev3](https://cloud.githubusercontent.com/assets/8686739/26191788/e0cdfba4-3be1-11e7-82aa-775dda3e3cda.PNG)
![es6](https://cloud.githubusercontent.com/assets/8686739/26191791/e0decc5e-3be1-11e7-8691-62680d670083.PNG)
![prod1](https://cloud.githubusercontent.com/assets/8686739/26191793/e0ead490-3be1-11e7-853f-1ab4a40ddec0.PNG)
![prod2](https://cloud.githubusercontent.com/assets/8686739/26191790/e0d342f8-3be1-11e7-8ef2-4058dd644ae4.PNG)
![command](https://cloud.githubusercontent.com/assets/8686739/26191925/67841502-3be2-11e7-9562-7152031cdfc3.PNG)
![webpack dev](https://cloud.githubusercontent.com/assets/8686739/26191942/78e98656-3be2-11e7-89d6-7612e5c4abb8.PNG)

@klimashkin @bebraw @emkay @marcus @jscheid , i also this such problem, please help us!

**Comment by krzksz** [comment]
Hey guys, maybe I'll be a little help for you.
@klimashkin Could you maybe share your webpack and babel configs? I just tried to run following code through webpack:
```javascript
if (process.env.NODE_ENV === 'production') {
  import('./something.js');
}
```
and it gets transpiled properly. Just make sure(and it's also a tip for @anlexN), that you have `babel-loader` included properly in module section in your configuration with `syntax-dynamic-import` plugin already mentioned above:
```javascript
 module: {
    rules: [
      {
        test: /\.js(x?)$/,
        exclude: /(node_modules)/,
        use: [
          {
            loader: 'babel-loader',
          },
        ],
      },
  ],
}
```
and my `.babelrc` config(you can also include this configuration right within `webpack.config.js`, see https://github.com/babel/babel-loader#options):
```javascript
{
  "presets": [
    [
      "env",
      {
        "modules": false,
        "targets": {
          "browsers": [
            "last 2 versions",
          ]
        }
      }
    ]
  ],
  "plugins": [
    "syntax-dynamic-import"
  ]
}
```

**Comment by klimashkin** [comment]
> and it gets transpiled properly

@krzksz Could you share what it's been transpiled to?

**Comment by michael-ciniawsky** [comment]
@klimashkin Are you using the `DefinePlugin` to set `process.env.NODE_ENV` ?
@anlexN If you use ES2015 Modules Syntax in your `webpack.config.js` rename it to `webpack.config.babel.js`

**Comment by klimashkin** [comment]
> Are you using the DefinePlugin to set process.env.NODE_ENV ?
@michael-ciniawsky Sure thats how 
```javascript
if (process.env.NODE_ENV === 'testing') {
  import('somemodule');
}
```
can be transformed to
```javascript
if (false) {
  import('somemodule');
}
```
If you read first message of this thread

**Comment by michael-ciniawsky** [comment]
? 🙃 Can't see any reference to `DefinePlugin` there. webpack may replace it with `false` by default if it isn't set, I can't tell with the current info provided

**Comment by klimashkin** [comment]
@michael-ciniawsky webpack doesn't replace `process.env.NODE_ENV` without `DefinePlugin`, it imports `node-libs-browser/~/process` in that case

**Comment by krzksz** [comment]
It would be best if you could provide the repository that recreates the issue so we can look deeply into that.

For me additional async chunk is getting created and webpack just generates it's code for loading it.

**Comment by anlexN** [comment]
@michael-ciniawsky, i don't use babel anything! why  rename webpack.config.js to webpack.config.babel.js?

**Comment by krzksz** [comment]
@anlexN It is only needed if you want to use newer syntax and language future in your configuration. See https://webpack.js.org/configuration/configuration-languages/ for more info.

**Comment by realseanp** [comment]
Any updates on this?  I am having this issue as well with the following code:

```js
if (module.hot) {
  module.hot.accept('./sagas', () => {
    // following line is the culprit
    import('./sagas').then(({ default: newSagas }) => {
      const promises = sagaTasks.map(task => {
        task.cancel();
        return task.done;
      });
      Promise.all(promises).then(() => {
        sagaTasks = newSagas.map(saga => sagaMiddleware.run(saga));
      });
    });
  });
}
```

I've tried using `syntax-dynamic-import` to fix the issue, but it persists even with that plugin. However, the issue is fixed when I use UglifyJSPlugin

**Comment by anlexN** [comment]
I don't like these: syntax-dynamic-import, webpack.config.babel.js, why? because it make no differences to directly use babel-loader. it also confuse many people like me to have to use babel, in the reality, it like a bug ! not about functional bug, surely about syntactical bug! Because i did read in the school, don't you are taught about what is systactic problem by your beautiful teacher?

**Comment by Ephem** [comment]
I had this same problem (also with module.hot that was turned into false at transpile-time, creating an unreachable block).

I managed to work around it by including something that could not be evaluated to false at transpile-time, forcing webpack to transpile the contents of the if-block:

```js
var alwaysFalse = false;
if (module.hot || alwaysFalse) {
  import('./SomeFile').then(() => {}); // Now this gets transpiled correctly
}
```

Maybe this can help someone else while this gets fixed.

**Comment by leidegre** [comment]
Running into this issue as well with latest version of webpack, a reasonable solution would be to use the uglify plugin to just strip the false conditional blocks because since that is what you meant to do anyway, however, there doesn't appear to be a straight forward way to do this without having your dev source mangled a bit and UglifyJs does add to the compilation time as well.

**Comment by bradennapier** [comment]
This completely breaks the entire concept of webpack.

This is 100% broken.  

Because I use `if (module.hot)` my app does not work in production...so i am going in and manually deleting all that code everywhere each time i want to go to production... stupid as fuck lolz

**Comment by alexlur** [comment]
I encountered the exact same problem: `import()` in unreachable blocks are not transpiled. If you want a quick fix, use `babel-plugin-transform-node-env-inline` to inject `NODE_ENV` in the meantime.

**Comment by anlexN** [comment]
Don't! It's painful!. This is webpack authors' wrong, not our developers' wrong......

**Comment by VitorLuizC** [comment]
Babel **should** at least transpile dynamic import on unreachable blocks. This should solve that problem without this painful.

**Comment by ooflorent** [comment]
Fixed in `4.0.0-alpha.5`.

**Comment by zemochen** [comment]
today I found the error appear again，in my old project use `html-webpack-plugin@4.0.0-alpha` it can worked,

in my project I config the `package.json` with `    "html-webpack-plugin": "^4.0.0-alpha",`

then the html-webpack-plugin update to `html-webpack-plugin@4.0.0-beta.1` it dosn't work
